File: src/org/lineageos/xiaomi_bluetooth/utils/PreferenceUtils.java
package org.lineageos.xiaomi_bluetooth.utils;

import android.content.Context;
import android.content.res.XmlResourceParser;

import androidx.annotation.NonNull;
import androidx.annotation.XmlRes;

import org.lineageos.xiaomi_bluetooth.settings.configs.ConfigController;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

public class PreferenceUtils {

    private static final String NAMESPACE_ANDROID = "http://schemas.android.com/apk/res/android";
    private static final String NAMESPACE_APP = "http://schemas.android.com/apk/res-auto";

    private static final String NAME_KEY = "key";
    private static final String NAME_CONTROLLER = "controller";

    private static final String PREFERENCE_TAG_SUFFIX = "Preference";

    @NonNull
    public static Set<ConfigController> createAllControllers(@NonNull Context context,
                                                             @XmlRes int xmlResId) {
        Set<ConfigController> configControllers = new HashSet<>();

        try (XmlResourceParser parser = context.getResources().getXml(xmlResId)) {
            int eventType;
            while ((eventType = parser.next()) != XmlPullParser.END_DOCUMENT) {
                if (eventType != XmlPullParser.START_TAG) {
                    continue;
                }
                String tagName = parser.getName();
                String key = parser.getAttributeValue(NAMESPACE_ANDROID, NAME_KEY);
                String controller = parser.getAttributeValue(NAMESPACE_APP, NAME_CONTROLLER);

                if (isPreferenceTag(tagName) && controller != null && key != null) {
                    configControllers.add(
                            createControllerInstance(context, controller, key));
                }
            }
        } catch (XmlPullParserException | IOException | ReflectiveOperationException e) {
            throw new RuntimeException("Error parsing XML or instantiating controller", e);
        }

        return configControllers;
    }

    private static boolean isPreferenceTag(@NonNull String tagName) {
        return tagName.endsWith(PREFERENCE_TAG_SUFFIX);
    }

    @NonNull
    private static ConfigController createControllerInstance(
            @NonNull Context context, @NonNull String controllerClass, @NonNull String preferenceKey
    ) throws ReflectiveOperationException {
        Object instance = Class.forName(controllerClass)
                .getConstructor(Context.class, String.class)
                .newInstance(context, preferenceKey);

        if (instance instanceof ConfigController controller) {
            return controller;
        } else {
            throw new IllegalStateException(controllerClass + " is not valid ConfigController");
        }
    }

}

File: src/org/lineageos/xiaomi_tws/utils/PreferenceUtils.kt
package org.lineageos.xiaomi_tws.utils

import android.bluetooth.BluetoothDevice
import android.content.Context
import androidx.annotation.XmlRes
import androidx.preference.Preference
import org.lineageos.xiaomi_tws.configs.BaseConfigController
import org.xmlpull.v1.XmlPullParser

object PreferenceUtils {

    private const val NAMESPACE_ANDROID = "http://schemas.android.com/apk/res/android"
    private const val NAMESPACE_APP = "http://schemas.android.com/apk/res-auto"

    private const val NAME_KEY = "key"
    private const val NAME_CONTROLLER = "controller"

    fun createAllControllers(
        context: Context,
        @XmlRes xmlResId: Int,
        device: BluetoothDevice
    ): Set<BaseConfigController<Preference>> {
        val configControllers = HashSet<BaseConfigController<Preference>>()

        runCatching {
            context.resources.getXml(xmlResId).use { parser ->
                var eventType: Int
                while ((parser.next().also { eventType = it }) != XmlPullParser.END_DOCUMENT) {
                    if (eventType != XmlPullParser.START_TAG) {
                        continue
                    }
                    val key = parser.getAttributeValue(NAMESPACE_ANDROID, NAME_KEY) ?: continue
                    val controller =
                        parser.getAttributeValue(NAMESPACE_APP, NAME_CONTROLLER) ?: continue

                    configControllers.add(createControllerInstance(controller, key, device))
                }
            }

            configControllers
        }.onFailure {
            throw RuntimeException("Error parsing XML or instantiating controller", it)
        }

        return configControllers
    }

    private fun createControllerInstance(
        controllerClass: String,
        preferenceKey: String,
        device: BluetoothDevice
    ): BaseConfigController<Preference> {
        val instance = Class.forName(controllerClass)
            .getConstructor(String::class.java, BluetoothDevice::class.java)
            .newInstance(preferenceKey, device)

        @Suppress("UNCHECKED_CAST")
        return instance as BaseConfigController<Preference>
    }

}
